baldrick/plugins/github_pull_requests_autolabel.py

```diff
@@ -0,0 +1,103 @@
+from os import path, sep
+
+from baldrick.github.github_api import RepoHandler
+from baldrick.plugins.github_pull_requests import pull_request_handler
+
+
+def get_subpackage_labels(files, all_labels):
+    """Loop through all modified file paths and extract the path elements for
+    each of the files. Then see if any of these paths or subsets of the path
+    exist as labels in the repo. This ignores the root of the path, so that,
+    e.g., astropy/coordinates and docs/coordinates both get caught and matched
+    to a "coordinates" label. Because this also considers all cumulative subsets
+    of the path, the following all map to, e.g., a "coordinates" label:
+
+        astropy/coordinates/file.py
+        astropy/coordinates/tests/file.py
+        astropy/coordinates/stuff/tests/file.py
+        docs/coordinates/stuff/file.rst
+
+    Parameters
+    ----------
+    files : iterable
+        A list or iterable of full path names to files modified by a pull
+        request.
+    all_labels : iterable
+        A list or iterable of all labels defined for the repository.
+
+    Returns
+    -------
+    labels : set
+        A set containing all subpackage labels to be added to the pull request
+        that already exist in the input ``all_labels``.
+
+    """
+    labels = set()
+
+    for file_ in files:
+        # Each file_ is assumed to be a full path to a modified file within
+        # the repository. This then grabs the directory name of the file_,
+        # stripping off the filename itself:
+        subdir = path.dirname(file_)
+
+        # If we are in a subdirectory:
+        if subdir:
+            # Get the subdirectory root name (e.g., "packagename" or "docs"),
```

Comment:
A package which follows the `src/` layout scheme might have multiple layers of root. I.e. you might have `docs/` and `src/astropy/`. I am not sure how is best to handle that.

Reply:
Me neither. Even in astropy core, it is not 100% correct when the subpackage is 3 levels deep. See astropy/astropy-bot#108 .